src/Analyzers/MSTest.Analyzers/DataRowShouldBeValidAnalyzer.cs

```diff
@@ -0,0 +1,214 @@
+            }
+
+            return;
+        }
+
+        // DataRow constructors have either zero or one argument(s). If we get here, we are
+        // on the one argument case. Check if we match either of the array argument constructors
+        // and expand the array argument if we do.
+        ImmutableArray<TypedConstant> constructorArguments = attribute.ConstructorArguments;
+        if (constructorArguments[0].Kind is TypedConstantKind.Array && !constructorArguments[0].IsNull)
+        {
+            constructorArguments = constructorArguments[0].Values;
+        }
+
+        if (IsArgumentCountMismatch(constructorArguments.Length, methodSymbol.Parameters.Length, lastMethodParameterIsArray))
+        {
+            context.ReportDiagnostic(syntax.CreateDiagnostic(
+                ArgumentCountMismatchRule,
+                constructorArguments.Length,
+                methodSymbol.Parameters.Length));
+            return;
+        }
+
+        // Check constructor argument types match method parameter types.
+        List<(int ConstructorArgumentIndex, int MethodParameterIndex)> typeMismatchIndices = new();
+        for (int i = 0, j = 0; i < constructorArguments.Length; ++i, ++j)
+        {
+            if (constructorArguments[i].IsNull)
+            {
+                continue;
+            }
+
+            ITypeSymbol? argumentType = constructorArguments[i].Type;
+            ITypeSymbol paramType = (lastMethodParameterIsArray && j >= methodSymbol.Parameters.Length - 1)
+                ? ((IArrayTypeSymbol)lastMethodParameter.Type).ElementType
+                : methodSymbol.Parameters[j].Type;
+
+            if (argumentType is not null && !argumentType.IsAssignableTo(paramType, context.Compilation))
+            {
+                typeMismatchIndices.Add((i, j));
+            }
+        }
+
+        // Report diagnostics if there's any type mismatch.
+        if (typeMismatchIndices.Count > 0)
+        {
+            context.ReportDiagnostic(syntax.CreateDiagnostic(
+                ArgumentTypeMismatchRule,
+                FormatTypeMismatchIndexList(typeMismatchIndices)));
+        }
+    }
+
+    private static bool IsArgumentCountMismatch(int constructorArgumentsLength, int methodParametersLength, bool lastMethodParameterIsArray)
+    {
+        // 1. If last method parameter is not an array the lengths must be the same.
+        // 2. If last method parameter is an array the argument count check is relaxed and we only
+        //    need to make sure we don't have less constructor arguments than actual method paramters.
+        return lastMethodParameterIsArray
+            ? constructorArgumentsLength < methodParametersLength - 1
+            : constructorArgumentsLength != methodParametersLength;
+    }
+
+    private static string FormatTypeMismatchIndexList(List<(int ConstructorArgumentIndex, int MethodParameterIndex)> typeMismatchIndices)
+    {
+        return string.Join(", ", typeMismatchIndices.ToArray());
```

Comment:
Join has an overload with `IEnumerable` so there is no need to pay for the `ToArray`.

Reply:
You're right, fixed it ! Thanks !